apps/passport-client/src/useSyncE2EEStorage.tsx

```diff
@@ -188,22 +181,134 @@ export type MergeStorageResult = APIResult<MergeableFields, NamedAPIError>;
+ * always returns the "remote" fields, making it equivalent to a simple
+ * download-and-replace.
+ *
+ * The merge performed here is limited by the lack of historical revisions, or
+ * other information which could clarify the user's intent.  E.g. we can't tell
+ * a remote "add" from a local "remove", and in case of replacement we don't
+ * know which version is newer.  Without that, we can't know which version of
+ * data is "better".  Instead we're opinionated on these tradeoffs:
+ * 1) Keeping data is always better than losing data.  If we can't tell if an
+ *    object was added or removed, assume it was added.  This means a conflict
+ *    may cause a removed object to reappear, but shouldn't cause a new object
+ *    to be lost.  This preference is chosen to avoid losing user data, but may
+ *    not always be ideal, e.g. in the case of subscriptions granting
+ *    permissions the user intended to revoke.
+ * 2) By default, prefer data downloaded from the server.  This is mostly
+ *    arbitrary, but given that we upload more aggressively than we download
+ *    it approximates a "first to reach the server wins" policy.  This means if
+ *    an object is modified (with the same ID) and involved in a conflict, one
+ *    of the two copies will be kept, determined by the timing of uploads.
+ *
+ * As a side-effect of the implementation of #2, this merge algorithm always
+ * replaces the PCDCollection and FeedsSubscriptionManager with new objects,
+ * discarding non-serialized state (like listeners on emitters, and
+ * subscription errors).
+ *
+ * This function always uploads a report to the server that a merge occurred,
+ * with some stats which we hope will be helpful to detect future problems
+ * and tune better merge algorithms.
  */
 export async function mergeStorage(
-  _localFields: MergeableFields,
+  localFields: MergeableFields,
   remoteFields: MergeableFields,
   self: User
 ): Promise<MergeStorageResult> {
-  console.error(
-    "[SYNC] sync conflict needs merge!  Keeping only remote state."
-  );
-  // TODO(artwyman): Refactor this out to implement and test real merge.
-  requestLogToServer(appConfig.zupassServer, "sync-merge", {
-    user: self.uuid
-    // TODO(artwyman): more details for tracking.
+  // Merge PCDs and Subscriptions independently, and gather a unified set of
+  // stats to include in a report to the server.
+  // TODO(#1372): Detect and report on cases where objects differ with the
+  // same ID.
+  let identical = true;
+  let anyPCDDiffs = false;
+  let anySubDiffs = false;
+
+  // PCD merge: Based on PCDCollection.merge, with predicate providing filtering
+  // by ID, and updating stats based on how many PCDs were added.
+  const pcdMergeStats = {
+    localOnly: 0,
+    remoteOnly: remoteFields.pcds.size() - localFields.pcds.size(),
+    both: localFields.pcds.size(),
+    final: remoteFields.pcds.size()
+  };
+  const pcds = remoteFields.pcds;
```

Comment:
The intent here was to conceptually separate the "result" collection from "local" or "remote.  That made more sense when I thought there might be a clone involved, or the algorithm might choose to keep either local or remote, but that ended up feeling unnecessary since the remote collection is always freshly deserialized anyway.
I'll just get rid of the extra variables.